py-polars/polars/utils/_construction.py

```diff
@@ -622,11 +622,12 @@ def _post_apply_columns(
         (columns or pydf_columns), schema_overrides=schema_overrides
     )
     column_subset: list[str] = []
-    if columns != pydf_columns:
-        if len(columns) < len(pydf_columns) and columns == pydf_columns[: len(columns)]:
-            column_subset = columns
-        else:
-            pydf.set_column_names(columns)
+    if set(pydf_columns).issuperset(columns):
```

Comment:
Could we come up with a cheaper check? One that doesn't require us to create a full set.

Reply:
Before this check, we just call `_unpack_schema` which may return a list no order fixed.

https://github.com/pola-rs/polars/blob/d3721f15167c080df6c843bb6d8b07105e1fc932/py-polars/polars/utils/_construction.py#L613-L623

https://github.com/pola-rs/polars/blob/d3721f15167c080df6c843bb6d8b07105e1fc932/py-polars/polars/utils/_construction.py#L651-L665

So the problem becomes that to check two series if one is supperset of the other which  one order not fixed.

We can sort both, then check if one is subslice of the other, the time complexity will be $O(nlogn + mlogm + min(m, n))$ at least(sort and compare), no more or a const  extra space required.

`set(pydf_columns).issuperset(columns)` or `set(columns).issubset(pydf_columns)` , the time complexity will be $O(m+ n)$ (convert and compare), $O(m)$ or $O(n)$ extra space required.

**I prefer _the less time_ way.**

we also can do some pre-check like this.
```python
    if len(columns) < len(pydf_columns) and set(pydf_columns).issuperset(columns):
        column_subset = columns
    elif len(columns) == len(pydf_columns):
        if set(pydf_columns).issuperset(columns):
            column_subset = columns
        else:
            pydf.set_column_names(columns)
    else:
        raise ShapeError(
            f"{len(columns)} column names provided "
            f"for a dataframe of width {len(pydf_columns)}"
        )
```

Reply:
I definitely want the pre-checks.

This is not by doing of you, but I didn't realize we can be so expensive after creating the `DataFrame`. I think we should look if we can hoist some work on to the Rust side, or elide that here. :thinking:

Reply:
I get it. I will see how to make it happen in Rust.

Reply:
> I get it. I will see how to make it happen in Rust.

Nice - but don't do anything yet; I think we can avoid doing anything in Python _without_ requiring any new Rust (for this case), if we accept losing one unusual/awkward override that we technically support at the moment ;)